packages/create-schemas/src/plugins/types-plugin.ts

```diff
@@ -0,0 +1,95 @@
+    return {
+        name: "internal:types-plugin",
+        async transform({ id, code }) {
+            if (id !== openapiTypeScriptId) {
+                return;
+            }
+
+            const ast = stringToAST(code) as ts.Node[];
+
+            const componentsDeclaration = ast.find(isComponentsInterfaceDeclaration);
+
+            assert(componentsDeclaration, "Missing components declaration");
+            const schema = componentsDeclaration.members.find(isComponentsSchema);
+            assert(schema, "Missing components declaration");
+            assert(schema.type && ts.isTypeLiteralNode(schema.type), "Invalid schema type");
+        
+            const typeNodes = schema.type.members
+                .map(member => member.name)
+                .filter(name => name !== undefined)
+                .filter(name => ts.isStringLiteral(name) || ts.isIdentifier(name))
+                .map(name => name.text)
+                .map(name => {
+                    if (RESERVED_IDENTIFIERS.has(name)) {
+                        throw new Error(`Invalid schema name: ${name}`);
+                    } else {
+                        return name;
+                    }
+                })
+                .map(name => `export type ${toSafeName(name)} = components["schemas"]["${name}"];`)
+                .flatMap(stringToAST) as ts.Node[];
+
+            const endpointsDeclaration = stringToAST("export type Endpoints = keyof paths;") as ts.Node[];
+
+            return {
+                code: code + "\n" + astToString([...typeNodes, ...endpointsDeclaration])
+            };
+        }
+    };
+}
+
+export function isComponentsInterfaceDeclaration(node: ts.Node): node is ts.InterfaceDeclaration {
+    return ts.isInterfaceDeclaration(node) && node.name.text === "components";
+}
+
+export function isComponentsSchema(node: ts.Node): node is ts.PropertySignature {
+    return ts.isPropertySignature(node) 
+    && (ts.isIdentifier(node.name) || ts.isStringLiteral(node.name))
+    && node.name.text === "schemas";
+}
+
+
+/**
+ * OpenAPI field names must match `^[a-zA-Z0-9\.\-_]+$` which allows names that
+ * are not valid JavaScript/TypeScript identifiers. This function converts an
+ * unsafe name into a safe name that can be used as a JavaScript/TypeScript
+ * identifier.
+ */
+export function toSafeName(unsafeName: string): string {
+    let safeName = "";
+    for (const char of unsafeName) {
+        const charCode = char.charCodeAt(0);
+
+        // A-Z
+        if (charCode >= 65 && charCode <= 90) {
+            safeName += char;
+        }
+
+        // a-z
+        if (charCode >= 97 && charCode <= 122) {
+            safeName += char;
+        }
+
+        if (char === "_" || char === "$") {
+            safeName += char;
+        }
+
+        // 0-9
+        if (safeName.length > 0 && charCode >= 48 && charCode <= 57) {
+            safeName += char;
+        }
+
+        continue;
+    }
+
+    return safeName;
+}
```

Comment:
In this case, I wonder if the approach should be have validation at spec generation to prevent the generation of names in the correct format instead. I think that should be a responsibility of the API side of things

Reply:
"Be conservative in what you do, be liberal in what you accept from others". 

I think the API-side should not produce ambiguous names, but the codegen tool should be able to handle them gracefully.